test/pagination/services.ts

```diff
@@ -0,0 +1,85 @@
+import { strict as assert } from 'node:assert';
+import { after, before, describe, it } from 'node:test';
+
+import * as dotenv from 'dotenv';
+dotenv.config({ path: './.env.test'});
+
+import { PrincipalService, recordToModel } from '../../src/principal/service.ts';
+import db, { insertAndGetId } from '../../src/database.ts';
+
+describe('findAll users pagination', () => {
+
+  const userRecordsDb: unknown[] = [];
+  before(async () => {
+
+    await db.schema.createTable('principals', (table) => {
+      table.increments('id').primary();
+      table.string('external_id').notNullable();
+      table.string('nickname').notNullable();
+      table.integer('type').notNullable();
+      table.boolean('active').notNullable();
+      table.timestamp('created_at').defaultTo(db.fn.now());
+      table.timestamp('modified_at').defaultTo(db.fn.now());
+      table.boolean('system').notNullable();
+    });
+
+    // 3 pages worth of users
+    for(let i = 1; i < 251; i++){
+      const data = {
+        external_id: i.toString(),
+        nickname: `User ${i}`,
+        type: 1,
+        active: 1,
+        created_at: Date.now(),
+        modified_at: Date.now(),
+        system: 0
+      };
```

Comment:
Also would it be possible to create these via `principalService.save()`

I'm a big fan of integration-style tests because you get more bang for your buck. More code gets tested, and the data that gets stored is more likely to be accurate with real usage.